File: atividade-14/atividade.go
package main

import "fmt"

type AFN struct {
	startState    string
	acceptStates  map[string]bool
	currentStates map[string]bool
}

func NewAFN() *AFN {
	return &AFN{
		startState:   "q0",
		acceptStates: map[string]bool{"q0": true, "q1": true, "q2": true},
		currentStates: map[string]bool{
			"q0": true,
		},
	}
}

func (afn *AFN) transition(input rune) {
	nextStates := make(map[string]bool)
	for state := range afn.currentStates {
		if state == "q0" {
			if input == '0' {
				nextStates["q1"] = true
			} else if input == '1' {
				nextStates["q2"] = true
			}
		} else if state == "q1" && input == '1' {
			nextStates["q2"] = true
		} else if state == "q2" && input == '0' {
			nextStates["q1"] = true
		}
	}
	afn.currentStates = nextStates
}

func (afn *AFN) IsAccepted(input string) bool {
	afn.currentStates = map[string]bool{
		afn.startState: true,
	}
	for _, c := range input {
		afn.transition(c)
	}

	return len(afn.currentStates) > 0
}

func main() {
	afn := NewAFN()
	fmt.Println(afn.IsAccepted("010101")) 
	fmt.Println(afn.IsAccepted("101010")) 
	fmt.Println(afn.IsAccepted("1101"))   
	fmt.Println(afn.IsAccepted("100"))    
	fmt.Println(afn.IsAccepted(""))       
}

File: atividade-14/atividade.py
class AFN:
    def __init__(self):
        self.states = {'q0', 'q1', 'q2'}  
        self.start_state = 'q0'
        self.accept_states = {'q0', 'q1', 'q2'}

    def is_accepted(self, string):
        current_states = {self.start_state}

        for char in string:
            next_states = set()
            for state in current_states:
                if state == 'q0':
                    if char == '0':
                        next_states.add('q1')  
                    elif char == '1':
                        next_states.add('q2')  
                elif state == 'q1':  
                    if char == '1':
                        next_states.add('q2')  
                elif state == 'q2':  
                    if char == '0':
                        next_states.add('q1')  
            current_states = next_states

        
        return len(current_states) > 0

if __name__ == '__main__':
    afn = AFN()
    print(afn.is_accepted("010101"))
    print(afn.is_accepted("101010"))
    print(afn.is_accepted("1101"))  
    print(afn.is_accepted("100"))   
    print(afn.is_accepted(""))